tools/PI/DevHome.PI/Helpers/ExternalTool.cs

```diff
@@ -113,44 +119,81 @@ private async void GetIcons()
-    }
-
-    internal string CreateCommandLine(int? pid, HWND? hwnd)
-    {
-        var commandLine = $" {OtherArgs}";
-
-        if (ArgType == ExternalToolArgType.Hwnd && hwnd is not null)
-        {
-            commandLine = $" {ArgPrefix} {hwnd:D} {OtherArgs}";
-        }
-        else if (ArgType == ExternalToolArgType.ProcessId && pid is not null)
-        {
-            commandLine = $" {ArgPrefix} {pid:D} {OtherArgs}";
-        }
-
-        return commandLine;
-    }
+    internal async Task<bool> Invoke(int? pid, HWND? hwnd)
+    {
+        var result = false;
 
-    internal virtual Process? Invoke(int? pid, HWND? hwnd)
-    {
         try
-        {
-            var toolProcess = new Process();
-            toolProcess.StartInfo.FileName = Executable;
-            toolProcess.StartInfo.Arguments = CreateCommandLine(pid, hwnd);
-            toolProcess.StartInfo.UseShellExecute = false;
-            toolProcess.StartInfo.RedirectStandardOutput = true;
-            toolProcess.Start();
-            return toolProcess;
+        {
+            var parsedArguments = string.Empty;
+            if (!string.IsNullOrEmpty(Arguments))
+            {
+                var argumentVariables = new Dictionary<string, int>();
+                if (pid.HasValue)
+                {
+                    argumentVariables.Add("pid", pid.Value);
+                }
+
+                if (hwnd.HasValue)
+                {
+                    argumentVariables.Add("hwnd", (int)hwnd.Value);
+                }
+
+                parsedArguments = ReplaceKnownVariables(Arguments, argumentVariables);
+            }
+
+            if (ActivationType == ToolActivationType.Protocol)
+            {
+                result = await Launcher.LaunchUriAsync(new Uri($"{parsedArguments}"));
+            }
+            else
+            {
+                if (ActivationType == ToolActivationType.Msix)
+                {
+                    var process = Process.Start("explorer.exe", $"shell:AppsFolder\\{AppUserModelId}");
+                    result = process is not null;
+                }
+                else
+                {
+                    var startInfo = new ProcessStartInfo(Executable)
+                    {
+                        Arguments = parsedArguments,
+                        UseShellExecute = true,
+                    };
+                    var process = Process.Start(startInfo);
+                    result = process is not null;
+                }
+            }
         }
         catch (Exception ex)
         {
             _log.Error(ex, "Tool launched failed");
-            return null;
-        }
+        }
+
+        return result;
+    }
+
+    private string ReplaceKnownVariables(string input, Dictionary<string, int> argumentValues)
+    {
+        // Process the input string to replace any instance of defined variables with "real" values.
+        // Eg, replace {pid} with 123, {hwnd} with 456.
+        var pattern = @"\{(\w+)\}";

```

Comment:
will `stupid` match with `pid`?

Reply:
No, the pattern looks for the curly braces.